### PR TITLE
chore(asm): iast oce request sampling

### PR DESCRIPTION
**Pull request by gnufede**
IAST: Use rate sampling in requests in order to reduce performance overhead caused by the IAST analysis.

## Checklist
- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/contributing.html#Release-Note-Guidelines) are followed.
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).
- [x] PR description includes explicit acknowledgement/acceptance of the performance implications of this PR as reported in the benchmarks PR comment.

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.
- [x] Reviewer has explicitly acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment.

### CODE REVIEW

**Review by smola — APPROVED** [review]

**Review by juanjux — APPROVED** [review]

**Review by avara1986 — DISMISSED** [review]

File: ddtrace/appsec/iast/_overhead_control_engine.py [hunk 1]
```diff
@@ -70,13 +78,17 @@ class OverheadControl(object):
...
+    _sampler = RateSampler(sample_rate=get_request_sampling_value() / 100.0)
+
+    def reconfigure(self):
+        self._sampler = RateSampler(sample_rate=get_request_sampling_value() / 100.0)
 
-    def acquire_request(self):
+    def acquire_request(self, span):  # type: (Span) -> None
         """Block a request's quota at start of the request.
 
         TODO: Implement sampling request in this method
```
Comment (Kyle-Verhoog) [inline ddtrace/appsec/iast/_overhead_control_engine.py:89]: stale TODO now?

File: riotfile.py [hunk 2]
```diff
@@ -226,6 +226,9 @@ def select_pys(min_version=MIN_PYTHON_VERSION, max_version=MAX_PYTHON_VERSION):
                 "cryptography": latest,
                 "astunparse": latest,
             },
+            env={
+                "DD_IAST_REQUEST_SAMPLING": "100",
```
[the remark this answers is not in the document]
Comment (gnufede) [reply riotfile.py]: Done!

File: ddtrace/appsec/iast/_overhead_control_engine.py [hunk 3]
```diff
@@ -9,15 +9,23 @@
...
     from typing import Tuple
     from typing import Type
 
+    from ddtrace.span import Span
+
 log = get_logger(__name__)
 
+
+def get_request_sampling_value():
+    return int(os.environ.get("DD_IAST_REQUEST_SAMPLING", 30))  # Percentage of requests analyzed by IAST
```
[the remark this answers is not in the document]
Comment (gnufede) [reply ddtrace/appsec/iast/_overhead_control_engine.py]: Thanks! Changed to float

File: ddtrace/appsec/iast/_overhead_control_engine.py [hunk 4]
```diff
@@ -70,13 +78,17 @@ class OverheadControl(object):
...
+        self._sampler = RateSampler(sample_rate=get_request_sampling_value() / 100.0)
 
-    def acquire_request(self):
+    def acquire_request(self, span):  # type: (Span) -> None
         """Block a request's quota at start of the request.
 
         TODO: Implement sampling request in this method
         """
-        if self._request_quota > 0:
+        if self._request_quota > 0 and self._sampler.sample(span):
```
[the remark this answers is not in the document]
Comment (gnufede) [reply ddtrace/appsec/iast/_overhead_control_engine.py:101]: I added some context in the docstring. The goal is to sample the IAST analysis, not the tracing, I double checked and IIUC adding this sampler here doesn't affect the sampling logic of the tracer, but correct me if I'm wrong

**Review by Kyle-Verhoog — APPROVED** [review]

**Review by avara1986 — APPROVED** [review]

**Review by juanjux — APPROVED** [review]

File: ddtrace/appsec/iast/_overhead_control_engine.py [hunk 5]
```diff
@@ -9,15 +9,24 @@
...
     from typing import Type
 
+    from ddtrace.span import Span
+
 log = get_logger(__name__)
 
+
+def get_request_sampling_value():  # type: () -> float
+    # Percentage of requests analyzed by IAST (default: 30%)
+    return float(os.environ.get("DD_IAST_REQUEST_SAMPLING", 30.0))
```
Comment (christophe-papazian) [inline ddtrace/appsec/iast/_overhead_control_engine.py:27]: Maybe add a try catch in case the variable is set to a non parsable float value like "foo" ?

**Review by christophe-papazian — APPROVED** [review]

File: ddtrace/appsec/iast/_overhead_control_engine.py (hunk 5 above)
Comment (gnufede) [reply ddtrace/appsec/iast/_overhead_control_engine.py:27]: Will do in a follow-up PR

**Pull request merged**